File: packages/typescript/goldenanalysis/src/core/analyzers/frameSummary.ts
/**
 * `frame.summary` — generic frame metrics, parity with the Python analyzer.
 *
 * Emits the Appendix-A metric set + a `per_column` table. NOTE: `frame.memory_bytes`
 * and the `per_column` `dtype` column are ENGINE-SPECIFIC (the Python sibling uses
 * polars `estimated_size()` / polars dtype names) and are emitted but are OUT of the
 * cross-surface parity contract — see the parity test + README.
 */

import {
  columns,
  duplicateRowRatio,
  nUnique,
  nullRatioPerColumn,
} from "../aggregate.js";
import type { Analyzer, AnalyzerInfo, AnalyzerInput, AnalyzerResult, AnalysisTable, FrameRows, Metric } from "../types.js";

const PRODUCES = [
  "frame.row_count",
  "frame.column_count",
  "frame.null_ratio_mean",
  "frame.duplicate_row_ratio",
  "frame.memory_bytes",
];

function dtypeLabel(rows: FrameRows, col: string): string {
  const kinds = new Set<string>();
  for (const row of rows) {
    const v = row[col];
    if (v === null || v === undefined) continue;
    kinds.add(typeof v);
  }
  if (kinds.size === 0) return "null";
  if (kinds.size > 1) return "mixed";
  return [...kinds][0]!;
}

export class FrameSummaryAnalyzer implements Analyzer {
  readonly info: AnalyzerInfo = {
    name: "frame.summary",
    consumes: ["frame"],
    produces: PRODUCES,
  };

  run(input: AnalyzerInput): AnalyzerResult {
    const rows: FrameRows = input.frame ?? [];
    const cols = columns(rows);
    const nRows = rows.length;
    const nCols = cols.length;
    const nullRatios = nullRatioPerColumn(rows, cols);
    const nullMean = nCols > 0 ? cols.reduce((acc, c) => acc + (nullRatios[c] ?? 0), 0) / nCols : 0;
    const dupRatio = duplicateRowRatio(rows, cols);
    // Portable byte estimate (NOT parity-asserted; the Python sibling uses polars).
    const memoryBytes = new TextEncoder().encode(JSON.stringify(rows)).length;

    const metrics: Metric[] = [
      { key: "frame.row_count", value: nRows, unit: "rows", direction: "neutral" },
      { key: "frame.column_count", value: nCols, unit: "columns", direction: "neutral" },
      { key: "frame.null_ratio_mean", value: nullMean, unit: "ratio", direction: "lower_better" },
      { key: "frame.duplicate_row_ratio", value: dupRatio, unit: "ratio", direction: "lower_better" },
      { key: "frame.memory_bytes", value: memoryBytes, unit: "bytes", direction: "neutral" },
    ];

    const perColumn: AnalysisTable = {
      name: "per_column",
      columns: ["column", "dtype", "null_ratio", "n_unique"],
      rows: cols.map((col) => [col, dtypeLabel(rows, col), nullRatios[col] ?? 0, nUnique(rows, col)]),
    };

    return { metrics, tables: [perColumn] };
  }
}

File: packages/python/goldenanalysis/goldenanalysis/analyzers/frame_summary.py
"""``frame.summary`` — generic frame metrics, zero suite deps, always available.

Emits the Appendix-A metric set: row/column counts, mean null ratio, exact-
duplicate-row ratio, estimated in-memory size, plus a ``per_column`` table
(column, dtype, null_ratio, n_unique).
"""

from __future__ import annotations

from goldenanalysis.core import aggregate as agg
from goldenanalysis.models import (
    AnalysisTable,
    AnalyzerInfo,
    AnalyzerInput,
    AnalyzerResult,
    Metric,
)

_PRODUCES = [
    "frame.row_count",
    "frame.column_count",
    "frame.null_ratio_mean",
    "frame.duplicate_row_ratio",
    "frame.memory_bytes",
]


class FrameSummaryAnalyzer:
    """Summarize a raw frame: shape, null mass, duplication, memory footprint."""

    info = AnalyzerInfo(name="frame.summary", consumes=["frame"], produces=_PRODUCES)

    def run(self, inp: AnalyzerInput) -> AnalyzerResult:
        df = inp.frame
        if df is None:
            raise ValueError("frame.summary requires AnalyzerInput.frame (a polars DataFrame)")

        n_rows = df.height
        n_cols = df.width
        null_ratios = agg.null_ratio_per_column(df)
        null_mean = sum(null_ratios.values()) / n_cols if n_cols else 0.0
        dup_ratio = agg.duplicate_row_ratio(df)
        mem_bytes = df.estimated_size()

        metrics = [
            Metric(key="frame.row_count", value=n_rows, unit="rows", direction="neutral"),
            Metric(key="frame.column_count", value=n_cols, unit="columns", direction="neutral"),
            Metric(
                key="frame.null_ratio_mean",
                value=null_mean,
                unit="ratio",
                direction="lower_better",
            ),
            Metric(
                key="frame.duplicate_row_ratio",
                value=dup_ratio,
                unit="ratio",
                direction="lower_better",
            ),
            Metric(key="frame.memory_bytes", value=mem_bytes, unit="bytes", direction="neutral"),
        ]

        per_column = AnalysisTable(
            name="per_column",
            columns=["column", "dtype", "null_ratio", "n_unique"],
            rows=[
                [col, str(df[col].dtype), null_ratios[col], df[col].n_unique()]
                for col in df.columns
            ],
        )

        return AnalyzerResult(metrics=metrics, tables=[per_column])
